File: Player.js
class Player extends Disc {
    constructor(centerX, centerY, radius, mass, restitution, damping, kickDamping, kickPower, color) {
        super(centerX, centerY, radius, mass, restitution, damping, color);
        this.agent = null;
        this.nonkickDamping = damping;
        this.kickDamping = kickDamping;
        this.kicker = new Kicker(this, kickPower);
    }

    applyAction(action) {
        if (action == Action.up) {
            this.applyForce(Vector.Unit.up);
            this.kicker.active = false;
        } else if (action == Action.upforward) {
            this.applyForce(this.agent.side == Side.red ? Vector.Unit.upright : Vector.Unit.upleft);
            this.kicker.active = false;
        } else if (action == Action.forward) {
            this.applyForce(this.agent.side == Side.red ? Vector.Unit.right : Vector.Unit.left);
            this.kicker.active = false;
        } else if (action == Action.downforward) {
            this.applyForce(this.agent.side == Side.red ? Vector.Unit.downright : Vector.Unit.downleft);
            this.kicker.active = false;
        } else if (action == Action.down) {
            this.applyForce(Vector.Unit.down);
            this.kicker.active = false;
        } else if (action == Action.downbackward) {
            this.applyForce(this.agent.side == Side.red ? Vector.Unit.downleft : Vector.Unit.downright);
            this.kicker.active = false;
        } else if (action == Action.backward) {
            this.applyForce(this.agent.side == Side.red ? Vector.Unit.left : Vector.Unit.right);
            this.kicker.active = false;
        } else if (action == Action.upbackward) {
            this.applyForce(this.agent.side == Side.red ? Vector.Unit.upleft : Vector.Unit.upright);
            this.kicker.active = false;
        } else if (action == Action.upshoot) {
            this.applyForce(Vector.Unit.up);
            this.kicker.active = true;
        } else if (action == Action.upforwardshoot) {
            this.applyForce(this.agent.side == Side.red ? Vector.Unit.upright : Vector.Unit.upleft);
            this.kicker.active = true;
        } else if (action == Action.forwardshoot) {
            this.applyForce(this.agent.side == Side.red ? Vector.Unit.right : Vector.Unit.left);
            this.kicker.active = true;
        } else if (action == Action.downforwardshoot) {
            this.applyForce(this.agent.side == Side.red ? Vector.Unit.downright : Vector.Unit.downleft);
            this.kicker.active = true;
        } else if (action == Action.downshoot) {
            this.applyForce(Vector.Unit.down);
            this.kicker.active = true;
        } else if (action == Action.downbackwardshoot) {
            this.applyForce(this.agent.side == Side.red ? Vector.Unit.downleft : Vector.Unit.downright);
            this.kicker.active = true;
        } else if (action == Action.backwardshoot) {
            this.applyForce(this.agent.side == Side.red ? Vector.Unit.left : Vector.Unit.right);
            this.kicker.active = true;
        } else if (action == Action.upbackwardshoot) {
            this.applyForce(this.agent.side == Side.red ? Vector.Unit.upleft : Vector.Unit.upright);
            this.kicker.active = true;
        } else if (action == Action.nomove) {
            this.kicker.active = false;
        } else if (action == Action.nomoveshoot) {
            this.kicker.active = true;
        }

    }



    applyActionHVS(actionH, actionV, actionS) {
        // TODO
        // HVS sirasında degilse exception yolla
        let diagonal = (actionH != ActionH.nomove) && (actionV != ActionV.nomove);
        let diagonalScale = diagonal ? Math.SQRT1_2 : 1;
        let shoot = (actionS == ActionS.shoot);

        if (actionH == ActionH.forward) {
            let forwardVec = this.agent.side == Side.red ? Vector.Unit.right : Vector.Unit.left;
            this.applyForce(Vector.mult(forwardVec, diagonalScale));
        } else if (actionH == ActionH.backward) {
            let backwardVec = this.agent.side == Side.red ? Vector.Unit.left : Vector.Unit.right;
            this.applyForce(Vector.mult(backwardVec, diagonalScale));
        }

        if (actionV == ActionV.up) {
            this.applyForce(Vector.mult(Vector.Unit.up, diagonalScale));
        } else if (actionV == ActionV.down) {
            this.applyForce(Vector.mult(Vector.Unit.down, diagonalScale));
        }

        if (actionS == ActionS.shoot) {
            this.kicker.activate();
        } else if (actionS == ActionS.nomove) {
            this.kicker.deactivate();
        }

    }




    update() {
        if (this.kicker.active) {
            this.damping = this.kickDamping;
        } else {
            this.damping = this.nonkickDamping;
        }
        super.update();
    }

    draw() {
        super.draw();
        this.kicker.draw();
    }
}

File: pyrl/Player.py
from Disc import Disc
from Kicker import Kicker
from Action import Action
from Vector import UnitVec
from Side import Side
from Color import Color


class Player(Disc):
    def __init__(self, center_x, center_y, radius, mass, restitution, damping, kick_damping, kick_power, side):
        if side == Side.red:
            color = Color.red
        elif side == Side.blue:
            color = Color.blue
        else:
            raise TypeError
        
        super().__init__(center_x, center_y, radius, mass, restitution, damping, color)
        self.side = side 
        self.nonkick_damping = damping
        self.kick_damping = kick_damping
        self.kicker = Kicker(self, kick_power)

    def apply_action(self, action):
        
        assert isinstance(action, Action)
        if action == Action.up:
            self.apply_force(UnitVec.up)
            self.kicker.deactivate()
            
        elif action == Action.upforward:
            self.apply_force(UnitVec.upright if self.side == Side.red else UnitVec.upleft)
            self.kicker.deactivate()
        
        elif action == Action.forward:
            self.apply_force(UnitVec.right if self.side == Side.red else UnitVec.left)
            self.kicker.deactivate()

        elif action == Action.downforward:
            self.apply_force(UnitVec.downright if self.side == Side.red else UnitVec.downleft)
            self.kicker.deactivate()

        elif action == Action.down:
            self.apply_force(UnitVec.down)
            self.kicker.deactivate()

        elif action == Action.downbackward:
            self.apply_force(UnitVec.downleft if self.side == Side.red else UnitVec.downright)
            self.kicker.deactivate()

        elif action == Action.backward:
            self.apply_force(UnitVec.left if self.side == Side.red else UnitVec.right)
            self.kicker.deactivate()

        elif action == Action.upbackward:
            self.apply_force(UnitVec.upleft if self.side == Side.red else UnitVec.upright)
            self.kicker.deactivate()

        elif action == Action.nomove:
            self.kicker.deactivate()
            
        elif action == Action.upshoot:
            self.apply_force(UnitVec.up)
            self.kicker.activate()
            
        elif action == Action.upforwardshoot:
            self.apply_force(UnitVec.upright if self.side == Side.red else UnitVec.upleft)
            self.kicker.activate()
        
        elif action == Action.forwardshoot:
            self.apply_force(UnitVec.right if self.side == Side.red else UnitVec.left)
            self.kicker.activate()

        elif action == Action.downforwardshoot:
            self.apply_force(UnitVec.downright if self.side == Side.red else UnitVec.downleft)
            self.kicker.activate()

        elif action == Action.downshoot:
            self.apply_force(UnitVec.down)
            self.kicker.activate()

        elif action == Action.downbackwardshoot:
            self.apply_force(UnitVec.downleft if self.side == Side.red else UnitVec.downright)
            self.kicker.activate()

        elif action == Action.backwardshoot:
            self.apply_force(UnitVec.left if self.side == Side.red else UnitVec.right)
            self.kicker.activate()

        elif action == Action.upbackwardshoot:
            self.apply_force(UnitVec.upleft if self.side == Side.red else UnitVec.upright)
            self.kicker.activate()
            
        elif action == Action.nomoveshoot:
            self.kicker.activate()

#        if action == Action.up:
#            self.apply_force(UnitVec.up)
#            self.kicker.deactivate()
#            
#        elif action == Action.upleft:
#            self.apply_force(UnitVec.upleft)
#            self.kicker.deactivate()
#        
#        elif action == Action.left:
#            self.apply_force(UnitVec.left)
#            self.kicker.deactivate()
#
#        elif action == Action.downleft:
#            self.apply_force(UnitVec.downleft)
#            self.kicker.deactivate()
#
#        elif action == Action.down:
#            self.apply_force(UnitVec.down)
#            self.kicker.deactivate()
#
#        elif action == Action.downright:
#            self.apply_force(UnitVec.downright)
#            self.kicker.deactivate()
#
#        elif action == Action.right:
#            self.apply_force(UnitVec.right)
#            self.kicker.deactivate()
#
#        elif action == Action.upright:
#            self.apply_force(UnitVec.upright)
#            self.kicker.deactivate()
#
#        elif action == Action.nomove:
#            self.kicker.deactivate()
#            
#        elif action == Action.upshoot:
#            self.apply_force(UnitVec.up)
#            self.kicker.activate()
#            
#        elif action == Action.upleftshoot:
#            self.apply_force(UnitVec.upleft)
#            self.kicker.activate()
#        
#        elif action == Action.leftshoot:
#            self.apply_force(UnitVec.left)
#            self.kicker.activate()
#
#        elif action == Action.downleftshoot:
#            self.apply_force(UnitVec.downleft)
#            self.kicker.activate()
#
#        elif action == Action.downshoot:
#            self.apply_force(UnitVec.down)
#            self.kicker.activate()
#
#        elif action == Action.downrightshoot:
#            self.apply_force(UnitVec.downright)
#            self.kicker.activate()
#
#        elif action == Action.rightshoot:
#            self.apply_force(UnitVec.right)
#            self.kicker.activate()
#
#        elif action == Action.uprightshoot:
#            self.apply_force(UnitVec.upright)
#            self.kicker.activate()
#            
#        elif action == Action.nomoveshoot:
#            self.kicker.activate()
#            
            
            


    def update(self):
        if self.kicker.active:
            self.damping = self.kick_damping
        else: 
            self.damping = self.nonkick_damping
        
        super().update()



    def draw(self):
        super().draw();
        self.kicker.draw()

    